### PR TITLE
[Part2] Partition and Sort Enforcement, ExecutionPlan enhancement

### LINKED ISSUE #189: Maintain partition information in Union
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-11991

Currently it returns Partitioning::UnknownPartitioning(num_partitions) based on the sum of partition counts (if available) of the underlying inputs. In case of another partition scheme - such as hash partitioning, it would be better to keep that information available, so other optimizations could be used.

### LINKED ISSUE #3400: output_partitioning() and output_ordering() implementations are wrong in some physical plan implementations with alias
**Describe the bug**

For example, in ProjectExec, output_partitioning and output_ordering should consider Alias and translate the child input accordingly.

**To Reproduce**

**Expected behavior**

**Additional context**

### LINKED ISSUE #3653: Revisit required_child_distribution(), output_partitioning(), output_ordering() implementations in ExecutionPlan's implementations
**Describe the bug**

Looks like couple of physical plan implementations do not implement those methods correctly.

**To Reproduce**

**Expected behavior**

**Additional context**

### LINKED ISSUE #3854: Partition and Sort Enforcement
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

In current DataFusion codebase, when it does the physical planning, it added RepartitionExec node blindly when it sees the hash join or aggregation without considering the children's output partitioning. It might add unnecessary RepartitionExecs
to the plan tree. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Partially Closes #3854.
Closes #3653
Closes #3400
Closes #189,

You can see the entire work in https://github.com/apache/arrow-datafusion/pull/3855

# Rationale for this change

# What changes are included in this PR?

1. Add methods required_input_ordering() to ExecutionPlan trait to specify the ordering requirements
2. Fix output_partitioning(), output_ordering(), required_input_distribution() in couple of trait implementations
3. Add method equivalence_properties() to ExecutionPlan trait to discover the equivalence properties in the Physical plan tree
4. Support partition aware UnionExec

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@alamb @andygrove @Dandandan @isidentical @yahoNanJing 
Please help to take a look

**Comment by alamb** [comment]
Thanks @mingmwang  -- I will review this carefully tomorrow

**Comment by alamb** [comment]
I am sorry -- I ran out of time today -- will try and find time tomorrow

**Comment by alamb** [comment]
@liukun4515 and @Ted-Jiang   perhaps you have some time to help review this as well

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -1515,4 +1515,84 @@ mod tests {
...
+    async fn partition_aware_union() -> Result<()> {
+        let left = test_table().await?.select_columns(&["c1", "c2"])?;
+        let right = test_table_with_name("c2")
+            .await?
+            .select_columns(&["c1", "c3"])?
+            .with_column_renamed("c2.c1", "c2_c1")?;
+
+        let left_rows = left.collect().await?;
+        let right_rows = right.collect().await?;
+        let join1 =
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1529]: Could you possibly add some comments here about what this test is verifying? It seems like perhaps it is verifying that when the joins are on the same key the partitioning is  the same and thus union can be done without bringing everything to a single stream?

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -1515,4 +1515,84 @@ mod tests {
...
+        let right_rows = right.collect().await?;
+        let join1 = left.join(
+            right.clone(),
+            JoinType::Inner,
+            &["c1", "c2"],
+            &["c2_c1", "c2_c2"],
+            None,
+        )?;
+
+        // join key ordering is different
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1570]: 👍

File: datafusion/core/src/physical_plan/joins/cross_join.rs [hunk 3]
```diff
@@ -153,16 +156,27 @@ impl ExecutionPlan for CrossJoinExec {
...
     fn output_partitioning(&self) -> Partitioning {
-        self.right.output_partitioning()
+        let left_columns_len = self.left.schema().fields.len();
+        adjust_right_output_partitioning(
+            self.right.output_partitioning(),
+            left_columns_len,
+        )
     }
 
+    // TODO check the output ordering of CrossJoin
```
Comment (alamb) [inline datafusion/core/src/physical_plan/joins/cross_join.rs:168]: is this still a todo?

File: datafusion/physical-expr/src/utils.rs [hunk 4]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
     list1.len() == list2.len() && list1.iter().zip(list2.iter()).all(|(e1, e2)| e1.eq(e2))
 }
 
+/// Assume the predicate is in the form of CNF, split the predicate to a Vec of PhysicalExprs.
+///
+/// For example, split "a1 = a2 AND b1 <= b2 AND c1 != c2" into ["a1 = a2", "b1 <= b2", "c1 != c2"]
+///
+pub fn split_predicate(predicate: &Arc<dyn PhysicalExpr>) -> Vec<&Arc<dyn PhysicalExpr>> {
```
Comment (alamb) [inline datafusion/physical-expr/src/utils.rs:81]: This is called `split_conjunction` in the logical optimizer -- perhaps it could be called the same thing in the physical layer. The logical expr implementation also avoids creating quite as many `Vec`s

https://github.com/apache/arrow-datafusion/blob/345234550712173477e7807ba2cf67dd2ffb9ed5/datafusion/optimizer/src/utils.rs#L58-L78

File: datafusion/physical-expr/src/utils.rs [hunk 5]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+            prop.insert(new_condition.1.clone());
+            idx1 = idx as i32;
+        } else if !contains_first && contains_second {
+            prop.insert(new_condition.0.clone());
+            idx2 = idx as i32;
+        } else if contains_first && contains_second {
+            idx1 = idx as i32;
+            idx2 = idx as i32;
+            break;
+        }
```
Comment (alamb) [inline datafusion/physical-expr/src/utils.rs:116]: You could also use a match statement here and let the compiler heck that all important cases are covered:

```suggestion
        match (contains_first, contains_second) {
            (true, false) => {
              prop.insert(new_condition.1.clone());
              idx1 = idx as i32;
            } 
            (false, true)=> {
              prop.insert(new_condition.0.clone());
              idx2 = idx as i32;
            }
            (true, true) =>  {
              idx1 = idx as i32;
              idx2 = idx as i32;
              break;
            }
            (false, false) => {}
        }
```

File: datafusion/physical-expr/src/utils.rs [hunk 6]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+    }
+}
+
+/// Combine the new equal condition with the existing equivalence properties.
+pub fn combine_equivalence_properties(
+    eq_properties: &mut Vec<EquivalenceProperties>,
+    new_condition: (&Column, &Column),
+) {
+    let mut idx1 = -1i32;
+    let mut idx2 = -1i32;
```
Comment (alamb) [inline datafusion/physical-expr/src/utils.rs:102]: I think typically in rust such a sentinel is signaled using `Option

So like

```rust
    let mut idx1: Option<usize> = None;
    let mut idx2: Option<usize> = None;
```

File: datafusion/physical-expr/src/physical_expr.rs [hunk 7]
```diff
@@ -136,6 +138,67 @@ impl PhysicalExprStats for BasicExpressionStats {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct EquivalenceProperties {
```
Comment (alamb) [inline datafusion/physical-expr/src/physical_expr.rs:142]: I wonder if `EquivalenceClass` is a more specific name?

Then you could make

```rust
struct EquivalenceProperties {
  classes: Vec<EquivalentClass>
}
```

And move functions like `truncate_equivalence_properties_not_in_schema` on to 

```rust
impl EquivalenceProperties {
  fn truncate_equivalence_properties_not_in_schema(&self, ..)
}
```

I don't think it is required, but it might keep the code easier to reason about / keep it behind an abstraction

File: datafusion/physical-expr/src/physical_expr.rs [hunk 8]
```diff
@@ -136,6 +138,67 @@ impl PhysicalExprStats for BasicExpressionStats {
     }
 }
 
+#[derive(Debug, Clone)]
```
Comment (alamb) [inline datafusion/physical-expr/src/physical_expr.rs:141]: What would you think about moving this into `datafusion/physical-expr/src/equivalence.rs` or something? Then we could move all the code that deals with equivalence classes into that module and keep them and the tests together

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 9]
```diff
@@ -186,13 +194,26 @@ impl AggregateExec {
 
         let schema = Arc::new(schema);
 
+        let mut alias_map: HashMap<Column, Vec<Column>> = HashMap::new();
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:197]: Can you explain what this code is for?  It doesn't seem correct to me as I don't understand the  circumstances under which the output of be different 🤔 

It seems like in this case the input logical plan maybe was incorrect?

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 10]
```diff
@@ -255,25 +276,58 @@ impl ExecutionPlan for AggregateExec {
...
+                    }
+                    _ => input_partition,
+                }
+            }
+            // Final Aggregation's output partitioning is the same as its real input
+            _ => self.input.output_partitioning(),
+        }
     }
 
+    // TODO check the output ordering of AggregateExec
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:305]: // is it still TODO?

File: datafusion/core/src/physical_plan/coalesce_batches.rs [hunk 11]
```diff
@@ -96,12 +96,15 @@ impl ExecutionPlan for CoalesceBatchesExec {
         self.input.output_partitioning()
     }
 
+    // Depends on how the CoalesceBatches was implemented, it is possible to keep
```
Comment (alamb) [inline datafusion/core/src/physical_plan/coalesce_batches.rs:99]: There is also `SortPreservingMerge` that can be used to preserve order but there are tradeoffs there (specifically it takes more effort to keep the sort order than it does to append batches together)

File: datafusion/core/src/physical_plan/filter.rs [hunk 12]
```diff
@@ -231,6 +246,38 @@ impl RecordBatchStream for FilterExecStream {
     }
 }
 
+/// Return the equals Column-Pairs and Non-equals Column-Pairs
+fn collect_columns_from_predicate(predicate: &Arc<dyn PhysicalExpr>) -> EqualAndNonEqual {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/filter.rs:250]: Perhaps this would be better in utils.rs

File: datafusion/physical-expr/src/physical_expr.rs (hunk 7 above)
Comment (alamb) [reply datafusion/physical-expr/src/physical_expr.rs:142]: Having `EquivalenceProperties` would also provide a single location to add docstrings explaining the structures, and their assumptions and what they are good for

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs [hunk 13]
```diff
@@ -119,22 +129,25 @@ impl ExecutionPlan for WindowAggExec {
...
-    fn required_child_distribution(&self) -> Distribution {
-        if self
-            .window_expr()
-            .iter()
-            .all(|expr| expr.partition_by().is_empty())
-        {
-            Distribution::SinglePartition
+    fn required_input_distribution(&self) -> Vec<Distribution> {
+        if self.partition_keys.is_empty() {
+            warn!("No partition defined for WindowAggExec!!!");
```
Comment (alamb) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:139]: I don't know why this would generate a warning -- can't this occur with a query like `SELECT ROW_NUMBER OVER () from foo` (as in an empty over clause)?

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs [hunk 14]
```diff
@@ -119,22 +129,25 @@ impl ExecutionPlan for WindowAggExec {
...
+            vec![Distribution::SinglePartition]
         } else {
-            Distribution::UnspecifiedDistribution
+            //TODO support PartitionCollections if there is no common partition columns in the window_expr
+            vec![Distribution::HashPartitioned(self.partition_keys.clone())]
```
Comment (alamb) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:143]: 👍  I agree this sounds good

File: datafusion/physical-expr/src/utils.rs [hunk 15]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+        })
+        .unwrap_or(expr)
+}
+
+pub fn normalize_expr_with_equivalence_properties(
+    expr: Arc<dyn PhysicalExpr>,
+    eq_properties: &[EquivalenceProperties],
+) -> Arc<dyn PhysicalExpr> {
+    let mut normalized = expr.clone();
+    if let Some(column) = expr.as_any().downcast_ref::<Column>() {
```
Comment (alamb) [inline datafusion/physical-expr/src/utils.rs:251]: Does this need to recursively rewrite exprs?

Like what if `expr` was `A + B` and you had an equivalence class with `B = C`

Wouldn't you have to rewrite `A + ` into `A + C`? But I don't see this code recursing.

This kind of rewrite could be tested as well I think

**Review by alamb — COMMENTED** [review]
Looking very impressive @mingmwang  -- thank you very much

My biggest question is how are the changes to distribution tested? I see code that verifies partitioning (or rather not partitioning) with UnionExec but there are changes made to all the other physical operators.

For example what about tests for `WindowAggregate` and outer joins and sort merge join?

I saw tests for some of the functions for operating on `EquivalenceProperties` 👍  but not all of them.

I left some style questions about encapsulating `EquivalenceProperties` that might also help


So TLDR is I think the changes to the physical operators need more tests.

Maybe you could break out the equivalence class code into a separate PR?

File: datafusion/physical-expr/src/physical_expr.rs (hunk 8 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/physical_expr.rs:141]: Sure

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs (hunk 13 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/windows/window_agg_exec.rs:139]: Yes, this is a valid case, but the SQL might run very slow without any "partition by" clause . I can remove the warning if we think the warning is useless.

File: datafusion/core/src/physical_plan/aggregates/mod.rs (hunk 10 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/aggregates/mod.rs:305]: Yes, I will remove it. The AggregateExec can not keep the ordering.

File: datafusion/core/src/physical_plan/aggregates/mod.rs (hunk 9 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/aggregates/mod.rs:197]: This is to deal with the case that there are alias in the group exprs, in this case we can not derive the output partitioning from the input/child directly, need to take the alias into consideration. This is similar to the `ProjectionExec`.

For example, the input has the output partitioning 'a', the `ProjectionExec` or `AggregateExec` might have alias 'a as a1', the output partitioning of the ProjectionExec or AggregateExec should be 'a1'.
`ProjectionExec` and `AggregateExec` will never change the real data distribution, but need to respect the alias.

File: datafusion/core/src/dataframe.rs (hunk 1 above)
Comment (mingmwang) [reply datafusion/core/src/dataframe.rs:1529]: Sure, this UT verifies the union is partition aware, by checking the output partitioning count is the default partition count(not the combinations of the input) and also the result is correct. I will add more comments here, and I will also add more code here to verify the output partition is same with the union's inputs.

File: datafusion/core/src/physical_plan/joins/cross_join.rs (hunk 3 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/joins/cross_join.rs:168]: Yeah, I'm not sure whether our CrossJoin implementation can keep the ordering of right side or not.

File: datafusion/physical-expr/src/utils.rs (hunk 4 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/utils.rs:81]: Sure.

File: datafusion/physical-expr/src/physical_expr.rs (hunk 7 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/physical_expr.rs:142]: Sure.

File: datafusion/physical-expr/src/utils.rs (hunk 6 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/utils.rs:102]: > Looking very impressive @mingmwang -- thank you very much
> 
> My biggest question is how are the changes to distribution tested? I see code that verifies partitioning (or rather not partitioning) with UnionExec but there are changes made to all the other physical operators.
> 
> For example what about tests for `WindowAggregate` and outer joins and sort merge join?
> 
> I saw tests for some of the functions for operating on `EquivalenceProperties` 👍 but not all of them.
> 
> I left some style questions about encapsulating `EquivalenceProperties` that might also help
> 
> So TLDR is I think the changes to the physical operators need more tests.
> 
> Maybe you could break out the equivalence class code into a separate PR?

In the 3rd Part of the PR, I will have 7~8 UTs to verify the distribution of the physical plans. Maybe they are more like
INTG tests, but the logic will be carefully verified.

File: datafusion/core/src/physical_plan/filter.rs (hunk 12 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/filter.rs:250]: Since this is only used by FilterExec, I would prefer to keep this as a private func in filter.rs

File: datafusion/physical-expr/src/physical_expr.rs (hunk 7 above)
Comment (yahoNanJing) [reply datafusion/physical-expr/src/physical_expr.rs:142]: Hi @alamb,  is there a kind of linked hash map in Rust? Maybe better to leverage a similar data structure here.

File: datafusion/physical-expr/src/utils.rs [hunk 16]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+                vec1
+            }
+            _ => vec![predicate],
+        },
+        None => vec![],
+    }
+}
+
+/// Combine the new equal condition with the existing equivalence properties.
+pub fn combine_equivalence_properties(
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/utils.rs:97]: Good interface design. It can be leveraged by both the Join and Filter

File: datafusion/physical-expr/src/utils.rs [hunk 17]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+    } else if idx1 == -1 && idx2 == -1 {
+        // adding new pairs
+        eq_properties.push(EquivalenceProperties::new(
+            new_condition.0.clone(),
+            vec![new_condition.1.clone()],
+        ))
+    }
+}
+
+pub fn remove_equivalence_properties(
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/utils.rs:138]: Why does the eq_properties contain the none equal columns?

File: datafusion/physical-expr/src/utils.rs [hunk 18]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+        if contains_first && contains_second {
+            match_idx = idx as i32;
+            break;
+        }
+    }
+    if match_idx >= 0 {
+        let matches = eq_properties.get_mut(match_idx as usize).unwrap();
+        matches.remove(remove_condition.0);
+        matches.remove(remove_condition.1);
+        if matches.len() <= 1 {
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/utils.rs:155]: This logic may be not correct. For example, original two equivalence properties, left side (l1,l2), right side (r1,r2), then after `combine_equivalence_properties`, it becomes one equivalence properties, (l1,l2,r1,r2). Then we comes to `remove_equivalence_properties` with remove condition (l1,r1).

File: datafusion/physical-expr/src/physical_expr.rs (hunk 7 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/physical_expr.rs:142]: The preferred data structure is LinkedHashSet.

File: datafusion/core/src/physical_plan/projection.rs [hunk 19]
```diff
@@ -51,6 +56,10 @@ pub struct ProjectionExec {
     schema: SchemaRef,
     /// The input plan
     input: Arc<dyn ExecutionPlan>,
+    /// The output ordering
+    output_ordering: Option<Vec<PhysicalSortExpr>>,
+    /// The alias map used to normalize out expressions like Partitioning and PhysicalSortExpr
+    alias_map: HashMap<Column, Vec<Column>>,
```
Comment (yahoNanJing) [inline datafusion/core/src/physical_plan/projection.rs:62]: Better to add comments to indicate what does the key & value stand for.

For my understanding, the key is the column in the input schema of this Projection operator. While the values are the columns in this output schema of this Projection operator.

File: datafusion/physical-expr/src/utils.rs [hunk 20]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+            eq_properties.remove(match_idx as usize);
+        }
+    }
+}
+
+pub fn merge_equivalence_properties_with_alias(
+    eq_properties: &mut Vec<EquivalenceProperties>,
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/utils.rs:162]: The `eq_properties` is the `EquivalenceProperties` of some input for the current operator.

Here, the goal of this function to construct a new `EquivalenceProperties` for the current operator

File: datafusion/physical-expr/src/utils.rs [hunk 21]
```diff
@@ -65,6 +74,210 @@ pub fn sort_expr_list_eq_strict_order(
...
+    alias_map: &HashMap<Column, Vec<Column>>,
+) {
+    for (column, columns) in alias_map {
+        let mut find_match = false;
+        for (_idx, prop) in eq_properties.iter_mut().enumerate() {
+            if prop.contains(column) {
+                for col in columns {
+                    prop.insert(col.clone());
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/utils.rs:170]: Although it can be corrected by `truncate_equivalence_properties_not_in_schema`, I still think it's better to construct a new one rather than do the merge based on the input `EquivalenceProperties`

File: datafusion/core/src/physical_plan/mod.rs [hunk 22]
```diff
@@ -122,10 +122,20 @@ pub trait ExecutionPlan: Debug + Send + Sync {
     /// have any particular output order here
     fn output_ordering(&self) -> Option<&[PhysicalSortExpr]>;
 
-    /// Specifies the data distribution requirements of all the
-    /// children for this operator
-    fn required_child_distribution(&self) -> Distribution {
-        Distribution::UnspecifiedDistribution
+    /// Specifies the data distribution requirements for all the
+    /// children for this operator, By default it's [[Distribution::UnspecifiedDistribution]] for each child,
+    fn required_input_distribution(&self) -> Vec<Distribution> {
```
Comment (yahoNanJing) [inline datafusion/core/src/physical_plan/mod.rs:127]: Maybe we can just use
`vec![Distribution::UnspecifiedDistribution; self.children().len()]`

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 23]
```diff
@@ -270,6 +271,75 @@ impl ExecutionPlan for HashJoinExec {
...
+                    .on
+                    .iter()
+                    .map(|(l, r)| {
+                        (
+                            Arc::new(l.clone()) as Arc<dyn PhysicalExpr>,
+                            Arc::new(r.clone()) as Arc<dyn PhysicalExpr>,
+                        )
+                    })
+                    .unzip();
+                vec![
```
Comment (yahoNanJing) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:291]: Currently it only supports exactly matched case. Is it possible to support partial matching case?

File: datafusion/physical-expr/src/utils.rs (hunk 18 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/utils.rs:155]: I agree it is confusing. I will remove the remove_equivalence_properties related logic.

File: datafusion/core/src/physical_plan/joins/hash_join.rs (hunk 23 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:291]: It is possible, but at current phase I will not implement it.

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 24]
```diff
@@ -270,6 +271,75 @@ impl ExecutionPlan for HashJoinExec {
...
+        match self.mode {
+            PartitionMode::CollectLeft => match self.join_type {
+                JoinType::Inner | JoinType::Right => adjust_right_output_partitioning(
+                    self.right.output_partitioning(),
+                    left_columns_len,
+                ),
+                JoinType::RightSemi | JoinType::RightAnti => {
+                    self.right.output_partitioning()
+                }
+                JoinType::Left
```
Comment (yahoNanJing) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:310]: Should these cases exist when the partition mode is CollectLeft?

File: datafusion/physical-expr/src/utils.rs (hunk 17 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/utils.rs:138]: I will remove the related logic.

File: datafusion/physical-expr/src/utils.rs (hunk 15 above)
Comment (mingmwang) [reply datafusion/physical-expr/src/utils.rs:251]: Yes, rewriting recursively is more safe.  Currently the equal join conditions are just Columns, 
and  for AggregateExec, the output_group_expr are also Columns. For WindowAggExec, does DataFusion support Partition by complex exprs ?

**Comment by mingmwang** [comment]
@alamb @yahoNanJing 
Please help to take look again.

**Comment by mingmwang** [comment]
retest please

File: datafusion/physical-expr/src/equivalence.rs [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+            for column in columns_to_remove {
+                class.remove(&column);
+            }
+        }
+        self.classes.retain(|props| props.len() > 1);
+    }
+}
+
+/// Equivalent Class is a set of Columns that are known to have the same value in all tuples in a relation
+/// Equivalence Class is generated by equality predicates, typically equijoin conditions and equality conditions in filters.
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/equivalence.rs:136]: I like this abstraction an the comments.

File: datafusion/physical-expr/src/equivalence.rs [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+
+    pub fn extend<I: IntoIterator<Item = EquivalentClass>>(&mut self, iter: I) {
+        self.classes.extend(iter)
+    }
+
+    /// Add new equal conditions into the EquivalenceProperties, the new equal conditions are usually comming from the
+    /// equality predicates in Join or Filter
+    pub fn add_equal_conditions(&mut self, new_conditions: (&Column, &Column)) {
+        let mut idx1: Option<usize> = None;
+        let mut idx2: Option<usize> = None;
```
Comment (yahoNanJing) [inline datafusion/physical-expr/src/equivalence.rs:48]: An option is much more correct now

File: datafusion/core/src/physical_plan/mod.rs [hunk 27]
```diff
@@ -472,7 +508,10 @@ pub enum Distribution {
     HashPartitioned(Vec<Arc<dyn PhysicalExpr>>),
```
Comment (yahoNanJing) [inline datafusion/core/src/physical_plan/mod.rs:508]: Maybe we need to add partition number and schema to the `HashPartitioned ` in the future.

**Review by yahoNanJing — APPROVED** [review]
LGTM

**Comment by mingmwang** [comment]
Should we make the Equivalence Properties schema aware ?

````
/// Equivalence Properties is a vec of EquivalentClass.
#[derive(Debug, Default, Clone)]
pub struct EquivalenceProperties {
    classes: Vec<EquivalentClass>,
    schema: SchemaRef,
}
````

**Comment by yahoNanJing** [comment]
> Should we make the Equivalence Properties schema aware ?

It would be great to add this schema constraint. Then we can avoid the ambiguous in 
https://github.com/apache/arrow-datafusion/blob/e945c37d25cb173d03929084bcd8aac31f71580e/datafusion/core/src/physical_plan/projection.rs#L202-L209

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs (hunk 13 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/windows/window_agg_exec.rs:139]: I would recommend removing the warning because it isn't clear to me what a user / administrator of the system would do in this case and so the warning will end up as spam in the logs I think. 

Perhaps we can just change it to `debug!`

File: datafusion/physical-expr/src/utils.rs (hunk 15 above)
Comment (alamb) [reply datafusion/physical-expr/src/utils.rs]: >  does DataFusion support Partition by complex exprs ?

Yes I think so:

```sql
DataFusion CLI v13.0.0
❯ create table foo as values (1,2), (3,4), (3,2), (2,1), (null, 0);

❯ select first_value(column1) over (partition by (column2%2) order by column2) from foo;
+--------------------------+
| FIRST_VALUE(foo.column1) |
+--------------------------+
| 2                        |
|                          |
|                          |
|                          |
|                          |
+--------------------------+
```

File: datafusion/core/src/physical_plan/mod.rs (hunk 22 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/mod.rs:127]: I personally think @yahoNanJing 's suggestion of

```rust
vec![Distribution::UnspecifiedDistribution; self.children().len()]
```

would make the intent clearer

File: datafusion/core/src/dataframe.rs [hunk 28]
```diff
@@ -1605,4 +1607,74 @@ mod tests {
...
+            match join_type {
+                JoinType::Inner
+                | JoinType::Left
+                | JoinType::LeftSemi
+                | JoinType::LeftAnti => {
+                    let left_exprs: Vec<Arc<dyn PhysicalExpr>> = vec![
+                        Arc::new(Column::new_with_schema("c1", &join_schema).unwrap()),
+                        Arc::new(Column::new_with_schema("c2", &join_schema).unwrap()),
+                    ];
+                    assert_eq!(
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1655]: 👍

File: datafusion/core/src/physical_plan/union.rs [hunk 29]
```diff
@@ -194,6 +258,73 @@ impl ExecutionPlan for UnionExec {
     }
 }
 
+/// CombinedRecordBatchStream can be used to combine a Vec of SendableRecordBatchStreams into one
```
Comment (alamb) [inline datafusion/core/src/physical_plan/union.rs:261]: I feel there was already a piece of code that does this -- maybe @tustvold  can remind me 🤔

**Review by alamb — APPROVED** [review]
Thanks @mingmwang  -- I agree this PR is ready to merge as is.

It would be great to file tickets to track your follow on work (like the TODO about cross joins, etc)

Thanks for getting this great stuff into DataFusion

**Comment by yahoNanJing** [comment]
Hi @alamb, should we merge this PR first so that @mingmwang will be able to continue the part 3 of this unnecessary shuffling optimization?

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs (hunk 13 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/windows/window_agg_exec.rs:139]: Sure

File: datafusion/core/src/physical_plan/mod.rs (hunk 22 above)
Comment (mingmwang) [reply datafusion/core/src/physical_plan/mod.rs:127]: Sure, I will change it in the following PR.

File: datafusion/core/src/dataframe.rs (hunk 28 above)
Comment (mingmwang) [reply datafusion/core/src/dataframe.rs:1655]: https://github.com/apache/arrow-datafusion/issues/4116
https://github.com/apache/arrow-datafusion/issues/4117
https://github.com/apache/arrow-datafusion/issues/4118

**Comment by alamb** [comment]
>  Hi @alamb, should we merge this PR first so that @mingmwang will be able to continue the part 3 of this unnecessary shuffling optimization?



Yes absolutely!

**Comment by alamb** [comment]
Sorry for the delay
